### PR TITLE
WebUI: Hook in to Polymer in order to add behaviors to components.

### PR DESCRIPTION
**Pull request by petemill**
Fix https://github.com/brave/brave-browser/issues/3631

Avoids having to patch each component and add the behaviors (and modules) each time.
Proof of concept using Settings page's Clear Browsing Data dialog behavior, but
can be applied to more modules in the future as settings (and other webui)
is optimized and modified.

Follow up from @mkarolin's https://github.com/brave/brave-core/pull/1213 which uses a minimizing-patch technique by adding a behavior to a polymer component in order to add new members to it, but as requested by @bridiver this further reduces the marginal effect of using this technique multiple times to zero by applying a patch once for each webui. This then provides a hook in to Polymer's module registration, which intercepts the module and modifies it before it is actually registered in Polymer.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [x] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
Ensure the new 'On Exit' tab shows up in the Settings -> Advanced -> Clear Browsing Data dialog.

## Reviewer Checklist:

- [x] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [x] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: browser/resources/settings/brave_settings_overrides.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+window.BravePatching.registerPolymerBehaviors({
+  'settings-clear-browsing-data-dialog': [
```
Comment (bsclifton) [inline browser/resources/settings/brave_settings_overrides.js]: IIRC, this name has to match the component name exactly, right?

Comment (petemill) [reply browser/resources/settings/brave_settings_overrides.js]: Yes @bsclifton. If I know what you're thinking, I'm guessing it's to validate that that is a name of a recognised chromium Component name? Unfortunately this function neccessarily gets called before the component is registered, which could happen at any time. There isn't a definitive time where we could say "OK, all the components have been registered, let's check that the expected ones still exist". Although we could try a DOM timing even, such as DOMContentLoaded, or even a manual debug check that happens after 10 seconds......?!

**Comment by petemill** [comment]
@bsclifton PR builder is cryptic as to its failure

**Comment by bsclifton** [comment]
Restarted PR builder after various fixes we've had - will keep an eye on it and merge if everything looks good 😄

**Comment by bsclifton** [comment]
Just rebased- it's going to kick off PR builder (build I did above failed because I only rebased `brave-browser` and not `brave-core`)

**Pull request merged**